[부산대 BE 이민형 - Step1] 엔티티 매핑

LeeSwallow opened the pull request:
## 특이사항

+ DataInitializer를 통해 admin 사용자와 필수적인 role들을 넣도록 수정했습니다. properties 속성을 통해 admin의 이메일과 비밀번호를 수정할 수 있도록 구성했습니다. -> 이는 서버 시작 시 `ServerStartVerifier`를 통해 검증됩니다.

+ `product`, `user`, `role`, `wished_product`에 대한 엔티티 매핑을 수행했습니다. 

+ 기존의 PK(product_id, user_id) 를 통해 M 대 N 매핑이었던 `wished_product`를 PK(id) 를 생성해 1:N 매핑으로 바꾸고 이를 연결하였습니다. 하지만 동일한 상품에 대한 생성은 충돌(409 Conflict) 처리 하였습니다. 단일 사용자가 다수의 위시리스트를 갖거나, 할인된 상품등에 대해 별도로 처리해야하는 상황이 있을 수 있지만 이는 추가 확장 시 변경해야할 문제로 생각했습니다.

+ role table은 PK(name) 의 단일 속성 테이블이므로 `user_roles` 테이블에서 PK(user_id, role_name) 으로 N:M 매핑으로 유지하였습니다.

+ userRepository에서 역할을 가져올 때, N+1 문제를  피하기 위해 `@EntityGraph`를 사용하였습니다. 

+ `wishedProductRepository` 또한 pagination 조회 시 상품 테이블을 `@EnitityGraph`를 통해 불러오도록 하여 N+1 문제를 회피하고자 했습니다. 

+ service Layer를 수정할 때 불필요한 null 검증(Controller에서 이미 검증한 속성) 등 일부를 수정하고 코드 재사용을 통해 간소화하였습니다.

+ JdbcClient로 구현하였던 `CustomPage` 객체를 재사용 하였습니다. JpaRepository의 Page 객체로 부터 Mapping 해오는 방식으로 CustomPage가 출력 됩니다.

+ `RestAssured` 를 통해 E2E Test를 수행하였습니다. RestDocs 설정을 해 수행과 동시에 문서를 생성합니다.  
+ 모든 test를 한번에 수행하면 병렬로 수행되는`@AfterEach`에 의해 충돌이 발생하는 문제를 발견하였습니다.`AFTER_EACH_TEST_METHOD` 콘텍스트를 부여해 순서대로 테스트를 수행하도록 하여 오류를 해결했습니다.
+ repostory unit 테스트를 구현했습니다.

### er 다이어그램

간단하게 mermaid로 er diagram을 그려보았습니다
<img width="758" height="892" alt="image" src="https://github.com/user-attachments/assets/614c92a0-cb17-4e94-a7ac-54170080f368" />


### 페이지네이션 구현

기본적으로 다음과 같이 조회할 대상과 페이지, 크기, 총 페이지, 총 요소 수가 부가 정보로 출력되도록 구현했습니다.

<img width="1906" height="1051" alt="image" src="https://github.com/user-attachments/assets/58940036-a0de-4eb0-8c9b-0342ca8945e3" />

wishlist를 조회할 때는 부가적으로 총 quantity 수와 가격 합계를 출력하도록 구성하였습니다.
이는 전체 요소에 대한 집계 결과 입니다 
<img width="1906" height="1044" alt="image" src="https://github.com/user-attachments/assets/e740ccbb-84e9-498f-a7b8-c4ac2294dd77" />


## step2 할 일 
+ 이전 단계에서 이미 CustomPage을 구현해놓아서 매핑만 시켜보았습니다. 
+ 이대로 진행해도 좋다면 `SortBy`를 구현하고 인메모리 Caching을 구현해볼 예정입니다. 대상은 조회가 빈번히 일어나는 Product가 될 것 같습니다.

## 질문사항
+ 페이지네이션을 구현할 때 Spring 표준 페이지 객체를 반환하는 것이 좋을까요?
+ `@EntityGraph`나 fetch 등을 쓰는 이유는 entity 객체의 확장성을 위해서 인가요? `WishedProduct`와 같이 product 정보가 꼭 필요한 객체는 Eager loading을 쓰는게 나을까요?
+ 순차적으로 테스트를 진행하니 속도가 2배 가량 느려졌습니다..병렬 테스팅 전략을 만들어보는게 나을까요?

Commit:
fix: 특정 엔티티가 삭제되기 전에 teardown 구문이 종료되는 오류 수정

- E2ETest 패키지로 전체 테스트 이관

Commit:
docs: 1 단계 구현 기능 목록 작성

Commit:
feat: jpa를 사용할 수 있도록 기본 설정 수정

- test 환경에서 동작 쿼리를 확인할 수 있도록 설정(`format_sql`=true, `show_sql`=true)
- `spring.jpa.hibernate.ddl-auto`=create-drop 설정
- h2에서 mysql dialect를 사용하도록 설정
- sql 스키마 수정(추후 삭제)

Commit:
fix: 특정 환경설정 미설정 시 오류가 나는 버그 수정

- GlobalExceptionHandler에서 Value를 받아올때 default = null로 하여 stacktrace 표시 미설정시 error response 에서 표시하지 않도록 수정

Commit:
feat: Product Entity & Repository 구현

- BaseEntity 에서 updatedAt과 createdAt 메타데이터를 받아오도록 설정
- jpaRepository 구현
- 기존의 dao 삭제

Commit:
fix: product service 로직과 controller 변경

- controller & service에서 patch endpoint 제거(개별 field값 수정은 jpa 단에서 최적화해주므로 무의미함)
- Page -> CustomPage로 mapping 하는 로직 추가
- Service가 CustomAuth 객체를 그대로 가져다 쓰는 함수 수정

Commit:
feat: ProductRepository에 대한 test 구성 완료

- sql scheme에 mysql 구문 사용 수정

Commit:
feat: User, Role entity & repository 구현

- Dao 삭제
- 연관 관계 & Lazy loading 설정
- EntityGraph를 통해 n+1 문제 해결

Commit:
feat: service & dto 수정

- find -> load 로 구현했던 로직을 fetch를 통해 한번에 조회하도록 수정
- pageRequest로 Page를 조회할 수 있도록 수정
- Dto Mapper 로직 수정

Commit:
feat: 서버 시작 시 관리자 계정과 역할을 추가하도록 구현

- DataInitializer 구현
- ServerStartupVerifier 에서 관련 설정을 검증 하도록 구현

Commit:
feat: wishList repository & entity 구현

- user, product를 ManyToOne으로 가져오도록 entity 설정
- WishedProductStat을 반환하도록 query 수정

Commit:
feat: wishList id를 만들고 해당 인텍스로 조회하도록 controller & service 수정

- (user_id, product_id) Pk -> wishlist_id 로 변경
- 관련 로직 수정

LeeSwallow committed:
fix: controller `@Valid` 누락 `@PathVariable` 필드 불일치 등의 에러 수정

LeeSwallow committed:
feat: Roles 파라미터에 대한 검증 validator 추가

LeeSwallow committed:
feat: E2E 테스트 변경사항에 맞게 수정

LeeSwallow committed:
feat: user,wishedProduct 에 대한 Repository 유닛Test 구현

urarik commented on src/main/java/gift/common/aop/GlobalExceptionHandler.java:
```diff
@@ -28,7 +28,7 @@ public class GlobalExceptionHandler {
     private Boolean showStackTrace = false;
 
-    public GlobalExceptionHandler(@Value("${server.error.include-stacktrace}")String stacktraceSetting) {
+    public GlobalExceptionHandler(@Value("${server.error.include-stacktrace:#{null}}")String stacktraceSetting) {
```
이런 옵션이 꼭 필요할까요?
스택 트레이스를 응답에 포함하는 것은 적절하지 않습니다.
만약 디버깅 등을 위해서 필요한 것이라면 서버측 로그에 남기는 것이 옳아요. 옵션이 필요한지 다시 한 번 생각해보세요~

LeeSwallow replied:
환경 설정에 대해 너무 파고들다 과하게 설정한것 같습니다. Spring 기본 에러처리 설정과 비슷하게 구현해보고 싶어 만들어보았습니다

urarik commented on src/main/java/gift/common/validation/validator/RoleListValidator.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.regex.Pattern;
+
+public class RoleListValidator implements ConstraintValidator<ValidRoleList, List<String>> {
+    Pattern rolePattern = Pattern.compile("^(ROLE_USER|ROLE_MD|ROLE_ADMIN)$");
```
역할이 추가될 때마다 해당 정규표현식을 수정해야 합니다. 효율적인 방법일까요?
코드를 수정한다는 것은 배포가 필요하다는 뜻입니다. 유연한 대응이 어려울 수 있어요

(추가) 다시 생각해보니 역할이 추가되거나 수정된다면 코드를 수정할 일이 다른 부분에서 어차피 꽤나 생길 것 같아요. 이 부분은 그대로 둬도 될 것 같습니다.

LeeSwallow replied:
서비스 단으로 검증 이관하고 role 존재여부를 판단하는 방식으로 수정하겠습니다

LeeSwallow replied:
UserRole.valueOf 으로 직접 파싱하는 방식을 사용하면 기존의 validator 를 활용하면서도 UserRole만을 의존하게 구성할 수 있습니다.

urarik commented on src/main/java/gift/config/DataInitializer.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Override
+    public void run(String... args) throws Exception {
+        roleRepository.saveAll(List.of
+                (new Role("ROLE_GUEST"),
+                new Role("ROLE_USER"),
+                new Role("ROLE_MD"),
+                new Role("ROLE_ADMIN")
+        ));
+        Role adminRole = roleRepository.findByName("ROLE_ADMIN").orElseThrow(
+                () -> new CriticalServerException("roleRepository가 제대로 작동하지 않습니다!")
+        );
+        User adminUser = new User();
+        adminUser.setEmail(adminEmail);
+        adminUser.setPassword(passwordEncoder.encode(adminPassword));
+        adminUser.setRoles(List.of(adminRole));
+        if (userRepository.existsByEmail(adminEmail)) {
+            throw new CriticalServerException("이미 사용 중인 관리자 이메일입니다: " + adminEmail);
+        }
+        userRepository.save(adminUser);
+    }
+}
```
제가 보기엔 data.sql과 properties로 더 간단하게 작성할 수 있을 것 같은데 이렇게 작성하신 이유가 있을까요~?

LeeSwallow replied:
앞의 코드와 같이 환경 설정을 활용한 기능 구현에 매몰되었던 것 같습니다. 
"시작 관리자 계정정보를 설정에 추가하는 편이 서버 관리자가 활용하기 더 편하지 않을까?" 라는 생각 흐름에서 초기 매니저 추가를 만들고 이후에 필수적으로 들어가야 하는 role도 추가하였습니다. 
지금 보니 너무 과하게 설정을 만든것 같긴 합니다

urarik commented on src/main/java/gift/config/ServerStartupVerifier.java:
```diff
@@ -16,25 +16,32 @@ public class ServerStartupVerifier {
    @PostConstruct
     public void verifyServerStartup() {
         log.info("서버 시작 검증을 시작합니다.");
         JwtVerification();
         passwordEncoderVerification();
+        adminCredentialsVerification();
```
여기 안에 있는 모든 검증 로직이 모두 필요한 로직일까요?
gift.jwt.secret을 사용하는 측에서 ${gift.jwt.secret:#{null}} 대신 ${gift.jwt.secret}으로 받으면 값이 없다는 것을 이런 추가적인 로직없이 알 수 있을텐데 이 코드들이 꼭 필요할까요~?

LeeSwallow replied:
서버 시작시 환경설정을 검증하고 추가적으로 코드를 확인하지 않고 로그만을 확인하여 문제점을 바로 수정할 수 있도록 구현하고 싶었습니다.

urarik replied:
@LeeSwallow 정답이 있는 문제가 아니니 이대로 하셔도 좋을 것 같아요. 
다만.. 제가 말씀드린 방법으로 해도 로그로 바로 문제가 보입니다. 즉, 로그만을 확인해서 문제점을 해결하기 위해 꼭 이 코드가 필요한건 아니예요. 그리고 제가 말씀드린 방법 또한 애플리케이션 시작 중에 감지할 수 있는 방법으로 감지 시점이 더 느리지도 않습니다. 설정을 실제로 사용하는 곳이 스택 트레이스로 찍히기 때문에 더 직관적이기도 하구요.
설정들이 늘어날수록 위 클래스에 로직들이 늘어날텐데 점점 유지보수하기 어려운 코드가 될 것 같아요. 설정이 늘어났는데 여기서 추가하지 않는다면? 안쓰는 설정을 사용하는 곳에서 지웠는데 여기서 안 지웠다면? 
제 생각엔  스프링에서 자동으로 해주는 일을 ServerStartupVerifier가 중복으로 하고 있는 것처럼 보여요. 길이 제한과 같은 검증 로직들은 있어도 될 것 같습니다만..

이렇게 쭉 말씀드리고 정답이 있는 문제가 아니라고 하면 받아들이기 어려우실 것 같긴하지만.. ㅎㅎㅎ; 제 생각이 틀릴 수도 있으니 그냥 이런 의견이 있다 정도로만 받아들여주세요.

urarik commented on src/main/java/gift/dto/user/UserCreateRequest.java:
```diff
@@ -16,20 +16,23 @@ public record UserCreateRequest(
         public User toEntity() {
-                Set<UserRole> roles = Set.of(UserRole.fromString(role));
-                return new User(null, email, password, roles);
+                var mappedRoles = roles.stream()
+                        .map(Role::new)
+                        .toList();
```
스트림 잘 사용해주셨네요 👍

urarik commented on src/main/java/gift/dto/wishlist/WishedProductResponse.java:
```diff
@@ -15,13 +17,17 @@ public record WishedProductResponse(
 ) {
     public static WishedProductResponse from(WishedProduct wishedProduct) {
+        Product product = wishedProduct.getProduct();
+        Long subtotal = product.getPrice() * wishedProduct.getQuantity();
```
이 로직의 적절한 위치가 dto일까요~? 정답은 없고 그냥 질문드리는거예요~

LeeSwallow replied:
dtoToEntity Mapper를 만들어서 관리할 수 있을 것 같습니다.

urarik replied:
@LeeSwallow 전 그냥 wishedProduct에서 getSubTotal()을 만드는 것을 생각했어요. 저 두 줄을 위해서 클래스를 또 하나 더 만드는 것은 불필요할 수 있는 복잡함을 추가하는 방법일 것 같습니다.

urarik commented on src/main/java/gift/entity/BaseEntity.java:
```diff
@@ -0,0 +1,38 @@
+    private Instant createdAt;
+    private Instant updatedAt;
+
+    @PrePersist
+    public void updateCreatedAt() {
+        if (this.createdAt == null) {
+            this.createdAt = Instant.now();
+        }
+        this.updatedAt = Instant.now();
+    }
+
+    @PreUpdate
+    public void updateUpdatedAt() {
+        this.updatedAt = Instant.now();
+    }
```
@CreatedDate, @LastModifiedDate 어노테이션을 공부해보세요~

urarik commented on src/main/java/gift/entity/Role.java:
```diff
@@ -0,0 +1,34 @@
+@Table(name = "roles")
+public class Role {
+    @Id
+    private String name;
```
UserRole 타입으로 하지 않고 String으로 하신 이유가 있나요~?

LeeSwallow replied:
Jpa에 익숙치 않아 `@Emumerate`이 생각나지 않았습니다

urarik commented on src/main/java/gift/repository/wishlist/WishedProductRepository.java:
```diff
@@ -1,18 +1,32 @@
+
+    Boolean existsByUserIdAndProductId(Long userId, Long productId);
+
+    @Query("""
+            SELECT new gift.entity.WishedProductStats(
+                SUM(w.quantity),
+                SUM(w.quantity * p.price)
+            )
+            FROM WishedProduct w
+            JOIN w.product p
+            WHERE w.user.id = :userId
+       """)
+    WishedProductStats calculateStatsByUserId(Long userId);
```
JPQL을 잘 다루시네요~ 👍

urarik commented on src/main/java/gift/service/auth/AuthServiceImpl.java:
```diff
@@ -5,81 +5,58 @@
-        User user = userRepository.findByEmail(email).orElseThrow(
-                () -> new UnauthorizedException("이메일 또는 비밀번호가 일치하지 않습니다.")
-        );
+        User user = userRepository.findByEmail(email)
+            .orElseThrow(() -> new UnauthorizedException("이메일 또는 비밀번호가 일치하지 않습니다."));
+
```
userService.getById 메소드가 이미 있는 것 같은데 이 메소드를 쓰면 중복된 예외 처리를 줄일 수 있지 않을까요~?

LeeSwallow replied:
UserService에 findByEmail 메소드를 만들고 AuthService에서 해당 메서드를 호출하는 걸 말씀하시는 걸까요?

urarik replied:
@LeeSwallow 아 그렇네요. 제가 헷갈렸습니다. findByEmail을 만들어서 사용해야겠네요. 그리고 다른 도메인의 repository를 직접적으로 사용하는 것은 그렇게 좋은 방법은 아니예요. 왜 그럴까요~?

urarik commented on src/main/java/gift/service/user/UserServiceImpl.java:
```diff
@@ -93,36 +55,31 @@ public User update(User user) {
         }
-
+        // 이메일 업데이트
         if (user.getEmail() != null) {
-            if (userRepository.findByEmail(user.getEmail()).isPresent()) {
+            if (userRepository.existsByEmail(user.getEmail())) {
                 throw new DuplicateKeyException("이미 존재하는 이메일입니다: " + user.getEmail());
             }
             existingUser.setEmail(user.getEmail());
         }
-
+        // 역할 업데이트
         if (user.getRoles() != null) {
             if (user.getRoles().isEmpty()) {
                 throw new IllegalArgumentException("사용자의 역할은 비어있을 수 없습니다.");
             }
-            if(!userRoleRepository.sync(user.getId(), user.getRoles())) {
-                throw new IllegalStateException("사용자의 역할을 업데이트하는데 실패했습니다. 사용자 ID: " + user.getId());
-            }
+            existingUser.setRoles(user.getRoles());
```
사용자가 이메일말고 다른 정보만 업데이트하려고 하면 해당 업데이트 로직이 성공할까요? 테스트 코드로 한 번 검증해보세요~

urarik commented on src/main/java/gift/service/wishlist/WishedProductServiceImpl.java:
```diff
@@ -29,115 +31,110 @@ public WishedProductServiceImpl(
     @Override
     @Transactional
-    public WishedProduct getByProductId(Long userId, Long productId) {
+    public WishedProduct getById(Long userId, Long wishedProductId) {
```
다른 메소드들을 볼 때도 공통적으로 드는 생각인데 repository는 findBy인데 왜 service에선 getBy인가요??

LeeSwallow replied:
repository와 메서드 명을 구분하는게 좋을 것 같았습니다. 수정하겠습니다.

urarik commented on src/main/java/gift/service/wishlist/WishedProductServiceImpl.java:
```diff
@@ -29,115 +31,110 @@ public WishedProductServiceImpl(
             throw new DuplicateKeyException("이미 장바구니에 존재하는 제품입니다. productId: " + productId);
         }
-        return wishedProductRepository.addProduct(userId, productId, quantity);
+        var productRef = productRepository.getReferenceById(productId);
+        var userRef = userRepository.getReferenceById(userId);
+        return wishedProductRepository.save(new WishedProduct(null, userRef, productRef, quantity));
```
getReferenceById 잘 써주셨네요~ 👍

urarik commented on src/test/java/gift/E2ETest/product/AbstractProductTest.java:
```diff
@@ -35,6 +36,18 @@ private UserAdminResponse getAdminResponse(UserCreateRequest request) {
                 .as(UserAdminResponse.class);
     }
 
+    private ProductDefaultResponse createProduct(ProductCreateRequest request) {
+        return RestAssured.given()
+                .header(AUTH_HEADER_KEY, this.adminToken)
+                .contentType("application/json")
+                .body(request)
+                .post(getRequestUrl())
+                .then()
+                .statusCode(201)
+                .extract()
+                .as(ProductDefaultResponse.class);
```
응답 코드를 검증하는 것도 좋지만 실제 응답 바디의 값을 검증하는 것도 필요하지 않을까요?

LeeSwallow replied:
해당 메서드는  추상 클래스에서 테스트 케이스 생성을 위해 정의해놓은 거라 검증은 따로 만들지 않았습니다.(테스트케이스 생성 도중 에러가 발생하는 상황을 체크하기 위해 status를 확인)
같은 엔드포인트에 대한 검증 및 문서화는 상속받는 테스트 클래스에서 수행하고 있습니다

LeeSwallow replied:
상속 구조는 다음과 같습니다.

AbstractControllerTest
│
├─ AbstractAuthTest
│      └─ AuthTest
│
├─ AbstractProductTest
│      ├─ ProductCreateTest
│      ├─ ProductDeleteTest
│      ├─ ProductReadTest
│      └─ ProductUpdateTest
│
├─ AbstractUserTest
│      ├─ UserCreateTest
│      ├─ UserDeleteTest
│      ├─ UserReadTest
│      └─ UserUpdateTest
│
└─ AbstractWishlistTest
       ├─ WishlistCreateTest
       ├─ WishListDeleteTest
       ├─ WishListReadTest
       └─ WishListUpdateTest

urarik commented on src/test/java/gift/UnitTest/repository/ProductRepositoryTest.java:
```diff
@@ -0,0 +1,85 @@
+        Product product = new Product(null, "Test Product", 1000L, "http://example.com/image.jpg",1L);
+        Product saved = productRepository.save(product);
+        assertAll (
+                () -> assertNotNull(saved.getId(), "상품 ID는 null이 아니어야 합니다."),
+                () -> assertEquals(product.getName(), saved.getName(), "상품 이름이 일치해야 합니다."),
+                () -> assertEquals(product.getPrice(), saved.getPrice(), "상품 가격이 일치해야 합니다."),
+                () -> assertEquals(product.getImageUrl(), saved.getImageUrl(), "상품 이미지 URL이 일치해야 합니다."),
+                () -> assertEquals(product.getOwnerId(), saved.getOwnerId(), "상품 소유자 ID가 일치해야 합니다.")
+        );
```
이미 assertEquals 등 메소드 이름에서 넣어주신 메시지의 의미가 내포되어 있습니다. 꼭 메시지를 다 추가해야할까요~? 불필요한 번거러운 작업처럼 보여요

LeeSwallow replied:
RestAssured 문서화 작업에 필요해서 습관적으로 넣었던것 같습니다 수정하겠습니다.

urarik reviewed (changes requested):
수고 많으셨습니다~ 아주 꼼꼼히 잘 작성해주셨네요. 코멘트 확인해주세요~ 하나 우려되는 점은.. 몇몇 부분들이 오버 엔지니어링처럼 보여요. 최대한 필요한 로직들을 간결하게 작성하는 것이 추후에 유지보수할 때 도움이 됩니다. 이 점만 염두에 두시면 될 것 같아요.

Step 2 할 일에선 SortBy 추가해보시는 것 좋은 방향인 것 같습니다. 다만 인메모리 caching은 수업 범위를 벗어나는 것 같아서 꼭 하실 필요는 없을 것 같습니다. 되게 열심히하시는게 눈에 보여서 혹시 시간이 남으신다면 JPA나 Java, Spring에 관련된 책을 한 권 읽어보시는게 어떨까요? 

질문사항에 대해서 말씀드리면,
1. Spring에서 제공하는 클래스를 사용할 수 있으면 사용하면 좋습니다. 그런데 이 클래스로 모자람이 있다면 새로운 클래스를 만들어야죠. 직접적인 연관은 없는 내용이지만 "effective java item 72" 에 대해 읽어보는 것도 좋을 것 같아요. 구글에 해당 키워드로 검색해보시면 글들이 나올거예요.
2. @EntityGraph나 fetch 전략을 적절하게 설정하는 이유는 성능 문제때문입니다. N+1 문제를 공부해보세요! 항상 가져와야하는 연관관계 엔티티인 경우엔 말씀하신 것처럼 EAGER를 쓰는 것이 제가 알기론 가장 깔끔합니다.
3. 테스트 속도는 @DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD) 가 추가된게 가장 큰 문제 같습니다. @SpringBootTest 를 사용할 때 성능에 가장 큰 영향을 미치는 것이 애플리케이션 컨텍스트를 만드는 부분입니다. 지금은 테스트 메소드가 끝날 때마다 새로 컨텍스트를 만들고 있으니 당연히 느릴 수 밖에 없어요. 우선 위 어노테이션이 꼭 필요한건지, 위 어노테이션을 쓰지 않고도 문제점을 해결할 수 있는지 확인해보시면 좋을 것 같아요~

LeeSwallow committed:
refactor: 불필요한 설정 기능 제거

- DataInitializer.java 제거
- 어드민 사용자 환경 설정 제거
- stacktrace 표시 관련 설정 제거
- sql을 통해 스키마를 초기화하도록 변경

LeeSwallow committed:
refactor: UserRole을 Id로 사용하도록 수정

- Enumerated 어노테이션을 활용해 UserRole을 키로 설정
- 기존의 String 기반 Role을 활용하던 로직 전반 수정

LeeSwallow committed:
refactor: test 로직 수정

- testuser testing 시 생성 & 삭제하여 독립적으로 테스팅되도록 수정
- WishedProductServiceImpl 로직 문제 해결

LeeSwallow committed:
refactor: validator 수정

LeeSwallow committed:
refactor: `CreateDate`, `LastModifiedDate`를 활용하도록 BaseEntity 변경

- entity 이전에 사용하던 AbstractEntity.java 삭제

LeeSwallow committed:
refactor: 일관적인 메서드 명을 가지도록 수정

LeeSwallow committed:
refactor: Mapper 구현 및 도입 완료

LeeSwallow committed:
refactor: 특정 필드 누락시 검증로직에 값 검증 추가

LeeSwallow commented:
### 수정 사항

+시작시 commandLineRunner로 data를 초기화하는 설정과 stacktrace를 on/off하는 설정 삭제
+ RoleList validation 로직 수정
+ createAt, updateAt 어노테이션을 통해 간단히 구현
+ 메서드명을 getBy -> findBy 로 통일
+ Mapper 구현

### 특이 사항

+ testcode에서 제대로 작동이 안되는 문제는 실제 wishedProductService 구현에 문제사항이 있었습니다.
+ EntityDtoMapper를 구현하였습니다.
+ ServerStartVerifier는 삭제 사항에서 제외하였습니다. 이유는 다음과 같습니다.
    1. 해당 클래스는 properties 관련 유효성 검증 또한 진행합니다(jwt 키 길이 등) 
    2. password encoding, jwt 관련 설정 관련 문제는 해당 로직이 수행 될 때 확인할 수 있습니다. 따라서 서버 시작 전에 미리 확인하고 문제가 있다면 서버 시작을 막는 것이 옳다고 느꼈습니다. 
    3. default null를 설정한 이유는 exception을 통해 문제를 확인하는 것보다 error log를 통해 확인하는 것이 문제 지점을 직관적으로 볼 수 있다고 (코드 확인 불필요) 생각했습니다.

LeeSwallow committed:
refactor: NamedEntityGraph를 활용해 재사용성 향상

LeeSwallow commented:
### 추가 구현
NamedEntityGraph를 추가해 재사용성을 증대했습니다.

urarik reviewed (approved):
수고 많으셨습니다! 추가적으로 대댓글을 달았는데 확인해주세요. 이와 별개로 코드 잘 작성해주셔서 머지하겠습니다.
한가지 말씀드리고 싶은 부분은 새로운 기능, 기술, 개념, 패러다임 등을 도입했을 때 시스템이 항상 더 좋아지는 것은 아닙니다. 어떤 방식이든 장단점이 있어요. 항상 장단점을 고려해보세요~
